### ISSUE TITLE
Features to implement in core

### ISSUE THREAD

**Issue by mdcfe** [Chaosthebot/Chaos-Core#1]
As per Chaosthebot/Chaos#565, the general consensus seems to be to use Node.js for the core, so I've started to add to branch `v2-pre-democracy` with an initial base to work from.

## Features to implement
- [ ] Self-updating from Chaos-Core repo
- [ ] Retrieving and updating components from specified repos
- [ ] Launching and restarting components from retrieved repos
- [ ] **Tests** (what frameworks?)
- Possibly others that I missed

@Chaosthebot/js-wizards

**Comment by andrewda** [comment]
As for tests, I love mocha and chai. Also istanbul/NYC for code coverage.

**Comment by PlasmaPower** [comment]
@andrewda Have you tried `jest`? It's like mocha+chai but better (multithreading tests for instance).

Edit: it also has an option builtin for code coverage.

**Comment by andrewda** [comment]
@PlasmaPower haven't used it but works for me!

**Comment by mdcfe** [comment]
I was about to suggest `jest`, mainly as it's one package and I think we should be trying to minimise our external dependencies. ~~Definitely not because I haven't used mocha/chai much, but using new tools is part of the challenge of different devs contributing to ChaosBot anyway imo~~

**Comment by Swizz** [comment]
And jest is all included : 
```sh
$ jest --coverage
```

**Comment by Redmega** [comment]
Agreed, `jest` is definitely the way to go.

As far as contributing to this, should we contribute to the `v2-pre-democracy` branch or should we create our own branches and open PRs to that one?

**Comment by andrewda** [comment]
Do we want to use some style guide like standard.js, Airbnb, etc.? Better to decide early.

**Comment by mdcfe** [comment]
@Redmega If you think the change might be controversial, open a PR, but if it's something simple or something that has already been agreed in an issue then I think it should be fine to commit directly to the branch.

**Comment by Swizz** [comment]
> Do we want to use some style guide like standard.js, Airbnb, etc.? Better to decide early.

I personnaly use [XO](https://github.com/sindresorhus/xo/blob/master/readme.md)

**Comment by mdcfe** [comment]
@andrewda I prefer Airbnb personally, but I'm open to a vote if someone compiles a list of presets to vote on

**Comment by Redmega** [comment]
I'd say neither. Use `prettier` and be done with it. Just add some configuration to allow dangling underscores. I love me some dangling underscores.

Edit: Just checked out `xo`, looks cool. I wouldn't be averse to using that either.

**Comment by Swizz** [comment]
Lets try [the new TJ things](https://github.com/tj/gh-polls).

---

[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/Prettier)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/Prettier/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/Airbnb)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/Airbnb/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/Standard%20)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/Standard%20/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/XO)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/XO/vote)

**click on an option to vote**

---

To add an option  : 
```md
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/**OPTION**)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9R8F6XQC1PN67RE2QV54EZ/**OPTION**/vote)
```

**Comment by andrewda** [comment]
@Swizz that's actually awesome! Airbnb and `prettier` sound like the best options to me. Voting for `prettier` just because it requires less setup.

**Comment by PlasmaPower** [comment]
As soon as I saw "new TJ thing" I knew it would be awesome :smile:

**Comment by mdcfe** [comment]
I'm not a fan of relying on code formatters personally - I'd much rather use ESLint to ensure that the code is kept consistent (`prettier-eslint`, though, looks quite appealing).

**Comment by Redmega** [comment]
Yeah that's the idea, that prettier and eslint can be used together to enforce during coding and format after coding.

What I usually do is have prettier run on filesave and have eslint police me during the coding.

**Comment by Swizz** [comment]
We could use [Husky](https://github.com/typicode/husky) to ensure the use of prettier before commit.

**Comment by mdcfe** [comment]
@Redmega @Swizz I think we could use `prettier-eslint-cli` with a git hook if we're going to use eslint alongside prettier to ensure that the eslint rules don't simply get overridden by prettier straight away. Husky could also be used to ensure that the tests pass before a commit (although this might be frustrating for some people).

**Comment by Swizz** [comment]
Add all the @Chaosthebot members into the loop

**Comment by andrewda** [comment]
We've got another important choice to make:

[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9SJDBM78ES5WG3TW43MJK3/Tabs)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9SJDBM78ES5WG3TW43MJK3/Tabs/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9SJDBM78ES5WG3TW43MJK3/Spaces)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BM9SJDBM78ES5WG3TW43MJK3/Spaces/vote)

We might want to also open a code formatting issue if this one's getting too big.
